File: app/src/main/kotlin/com/test/presentation/internal/Presenter.kt
package com.test.presentation.internal

import io.reactivex.Observable
import io.reactivex.disposables.Disposables
import io.reactivex.subjects.BehaviorSubject

abstract class Presenter<View : ViewStateRenderer<ViewState>, ViewState> {
  private val viewStates = BehaviorSubject.create<ViewState>()
  private var viewStateDisposable = Disposables.disposed()
  private var isActive = false

  abstract fun bindActions()

  fun attachView(view: View) {
    viewStateDisposable = viewStates.subscribe(view::render)
  }

  fun detachView() {
    viewStateDisposable.dispose()
  }

  private fun subscribeViewState(viewStateObservable: Observable<ViewState>) {
    check(!isActive) {
      "subscribeViewState() is only allowed to be called once"
    }
    isActive = true

    viewStateObservable.subscribe(viewStates)
  }

  @JvmName("_subscribeViewState")
  protected fun Observable<ViewState>.subscribeViewState() = subscribeViewState(this)
}

File: ios/Sources/ios/Presentation/Internal/Presenter.swift
import RxRelay
import RxSwift

open class Presenter<View: ViewStateRenderer> {
    typealias ViewState = View.ViewStateType
    var isBound = false
    private var subscribeViewStateMethodCalled = false
    private let viewStateRelay = BehaviorRelay<ViewState?>(value: nil)
    private var disposables: DisposeBag? = DisposeBag()

    init() {}

    open func bindActions() {
        fatalError("bindActions not implemented!. Add\n\npublic override func bindActions() {}\n\n")
    }

    final func internalBindActions() {
        bindActions()
        isBound = true
    }

    final func internalUnbind() {
        isBound = false
        disposables = nil
    }


    final func subscribeViewState(_ viewStateObservable: Observable<ViewState>) {
        guard !subscribeViewStateMethodCalled else {
            fatalError("subscribeViewState function is only allowed to be called once")
        }
        subscribeViewStateMethodCalled = true
        disposables?.insert(viewStateObservable.bind(to: viewStateRelay))
    }

    final func attachView(
        view: View,
        onViewState: @escaping (ViewState?) -> Void
    ) {
        disposables?.insert(viewStateRelay
            .observe(on: MainScheduler.instance)
            .subscribe(onNext: onViewState)
        )
    }
}

extension Observable {
    func subscribeViewState<PresenterImpl, View>(
        to presenter: PresenterImpl
    ) where PresenterImpl: Presenter<View>, View: ViewStateRenderer {
        guard let self = self as? Observable<PresenterImpl.ViewState> else {
            fatalError("subscribeViewState stream elements must conform to \(PresenterImpl.ViewState.self)")
        }
        presenter.subscribeViewState(self)
    }
}
